Python/pubsub/gcloud_logs_filter_with_dlq.py

```diff
@@ -0,0 +1,129 @@
+            "--output_topic",
+            default="projects/your-project/topics/your-output-test",
+            help="Output PubSub topic")
+        parser.add_argument(
+            "--dlq_topic",
+            default="projects/your-project/topics/your-dlq-test",
+            help="Dead Letter Queue PubSub topic")
+
+
+def run():
+    """
+    This Apache Beam pipeline processes log messages from a Google Cloud Pub/Sub topic.
+    The expected data format follows the standard Google Cloud log format,
+    which can be achieved by routing logs to a Pub/Sub topic via https://console.cloud.google.com/logs/router.
+
+    It performs the following steps:
+    1. Input Configuration:
+        - Reads messages from the specified input Pub/Sub topic.
+
+    2. Message Parsing:
+        - Parses each message as a JSON dictionary.
+
+    3. Message Splitting:
+        - Divides messages into two categories:
+        a. 'processed' (contain both 'severity' and 'jsonPayload' fields)
+        b. 'unprocessed' (missing one or both of these fields).
+
+    4. Severity Filtering:
+        - For 'processed' messages, filters out those with severity other than "ERROR".
+
+    5. Data Transformation:
+        - Extracts timestamp and message content from the 'jsonPayload' field for 'processed' messages.
+
+    6. Output Handling:
+        - Writes transformed 'processed' messages to a specified output Pub/Sub topic.
+        - Sends 'unprocessed' messages to a Dead Letter Queue (DLQ) topic.
+    """
+
+    options = PubSubOptions(streaming=True)
+
+    with beam.Pipeline(options=options) as p:
+        split_result = (p | "Read from PubSub" >> ReadFromPubSub(topic=options.input_topic)
+                            | "Parse JSON" >> Map(lambda msg: json.loads(msg))
+                            | "Split Messages" >> ParDo(SplitMessages()).with_outputs("unprocessed", "processed"))
+
+        # Filter processed messages and write to output topic
+        (split_result['processed']
+         | "Filter by Severity" >> Filter(filter_by_severity)
+         | "Transform Data" >> Map(transform_data)
+         | "Write to PubSub" >> WriteToPubSub(options.output_topic, with_attributes=True))
+
+        # Write unprocessed messages to DLQ
+        (split_result['unprocessed']
+         | "Map to PubsubMessage" >> Map(to_pubsub_message)
+         | "Write to DLQ" >> WriteToPubSub(options.dlq_topic, with_attributes=True))
+
+
+class SplitMessages(DoFn):
+    def process(self, element):
+        from apache_beam.pvalue import TaggedOutput
+
+        if ('severity' in element) & ('jsonPayload' in element):
+            yield TaggedOutput('processed', element)
+        else:
+            yield TaggedOutput('unprocessed', element)
+
+
+def filter_by_severity(log):
+    # Filter logs by severity level (only process logs with severity "ERROR")
+    return log.get("severity").upper() == "ERROR"
+
+
+def to_pubsub_message(msg):
+    from apache_beam.io import PubsubMessage
```

Comment:
Please move imports to the top of the file

Reply:
I used the same import as in the another PubSub pipeline in this repo, so I think it's ok.